miden-crypto/src/encryption/aead_rpo/mod.rs

```diff
@@ -0,0 +1,550 @@
+    // CONSTRUCTORS
+    // --------------------------------------------------------------------------------------------
+
+    /// Creates a new random secret key using the default random number generator.
+    #[cfg(feature = "std")]
+    #[allow(clippy::new_without_default)]
+    pub fn new() -> Self {
+        use rand::{SeedableRng, rngs::StdRng};
+        let mut rng = StdRng::from_os_rng();
+
+        Self::with_rng(&mut rng)
+    }
+
+    /// Creates a new random secret key using the provided random number generator.
+    pub fn with_rng<R: Rng>(rng: &mut R) -> Self {
+        rng.sample(StandardUniform)
+    }
+
+    // ENCRYPTION
+    // --------------------------------------------------------------------------------------------
+
+    /// Encrypts and authenticates the provided sequence of field elements using this secret key
+    /// and a random nonce.
+    #[cfg(feature = "std")]
+    pub fn encrypt_felts(&self, data: &[Felt]) -> Result<EncryptedData, EncryptionError> {
+        self.encrypt_felts_with_associated_data(data, &[])
+    }
+
+    /// Encrypts the provided sequence of field elements and authenticates both the ciphertext as
+    /// well as the provided associated data using this secret key and a random nonce.
+    #[cfg(feature = "std")]
+    pub fn encrypt_felts_with_associated_data(
+        &self,
+        data: &[Felt],
+        associated_data: &[Felt],
+    ) -> Result<EncryptedData, EncryptionError> {
+        use rand::{SeedableRng, rngs::StdRng};
+        let mut rng = StdRng::from_os_rng();
+        let nonce = Nonce::with_rng(&mut rng);
+
+        self.encrypt_felts_with_nonce(data, associated_data, nonce)
+    }
+
+    /// Encrypts and authenticates the provided data using this secret key and a random nonce.
+    ///
+    /// Before encryption, the bytestring is converted to a sequence of field elements.
+    #[cfg(feature = "std")]
+    pub fn encrypt(&self, data: &[u8]) -> Result<EncryptedData, EncryptionError> {
+        self.encrypt_with_associated_data(data, &[])
+    }
+
+    /// Encrypts the provided data and authenticates both the ciphertext as well as the provided
+    /// associated data using this secret key and a random nonce.
+    ///
+    /// Before encryption, both the data and the associated data are converted to sequences of
+    /// field elements.
+    #[cfg(feature = "std")]
+    pub fn encrypt_with_associated_data(
+        &self,
+        data: &[u8],
+        associated_data: &[u8],
+    ) -> Result<EncryptedData, EncryptionError> {
+        use rand::{SeedableRng, rngs::StdRng};
+        let mut rng = StdRng::from_os_rng();
+        let nonce = Nonce::with_rng(&mut rng);
+
+        self.encrypt_with_nonce(data, associated_data, nonce)
+    }
+
+    /// Encrypts the provided sequence of field elements and authenticates both the ciphertext as
+    /// well as the provided associated data using this secret key and the specified nonce.
+    pub fn encrypt_felts_with_nonce(
```

Comment:
Did we rename these methods to keep naming consistent with the non-AOE scheme? If so, it is not clear to me whether this is needed as this scheme is primarily for encrypting field elements - and so, there is an argument to be made that encrypting elements should be the default.

If we do go with this approach, should we go with "elements" instead of "felts"? (we usually use "elements" in method names).

Reply:
Let me think this through a bit. Maybe the right way is to add both sets of methods to both schemes. Basically, both schemes would have `encrypt_bytes()` and `encrypt_elements()` etc.